src/dbnode/storage/shard.go

```diff
@@ -274,32 +277,33 @@ func newDatabaseShard(
 		SubScope("dbshard")
 
 	s := &dbShard{
-		opts:                 opts,
-		seriesOpts:           seriesOpts,
-		nowFn:                opts.ClockOptions().NowFn(),
-		state:                dbShardStateOpen,
-		namespace:            namespaceMetadata,
-		shard:                shard,
-		namespaceReaderMgr:   namespaceReaderMgr,
-		increasingIndex:      increasingIndex,
-		seriesPool:           opts.DatabaseSeriesPool(),
-		reverseIndex:         reverseIndex,
-		lookup:               newShardMap(shardMapOptions{}),
-		list:                 list.New(),
-		newMergerFn:          fs.NewMerger,
-		newFSMergeWithMemFn:  newFSMergeWithMem,
-		filesetsFn:           fs.DataFiles,
-		filesetPathsBeforeFn: fs.DataFileSetsBefore,
-		deleteFilesFn:        fs.DeleteFiles,
-		snapshotFilesFn:      fs.SnapshotFiles,
-		sleepFn:              time.Sleep,
-		identifierPool:       opts.IdentifierPool(),
-		contextPool:          opts.ContextPool(),
-		flushState:           newShardFlushState(),
-		tickWg:               &sync.WaitGroup{},
-		coldWritesEnabled:    namespaceMetadata.Options().ColdWritesEnabled(),
-		logger:               opts.InstrumentOptions().Logger(),
-		metrics:              newDatabaseShardMetrics(shard, scope),
+		opts:                      opts,
+		seriesOpts:                seriesOpts,
+		nowFn:                     opts.ClockOptions().NowFn(),
+		state:                     dbShardStateOpen,
+		namespace:                 namespaceMetadata,
+		shard:                     shard,
+		namespaceReaderMgr:        namespaceReaderMgr,
+		increasingIndex:           increasingIndex,
+		seriesPool:                opts.DatabaseSeriesPool(),
+		reverseIndex:              reverseIndex,
+		lookup:                    newShardMap(shardMapOptions{}),
+		list:                      list.New(),
+		newMergerFn:               fs.NewMerger,
+		newFSMergeWithMemFn:       newFSMergeWithMem,
+		filesetsFn:                fs.DataFiles,
+		filesetPathsBeforeFn:      fs.DataFileSetsBefore,
+		deleteFilesFn:             fs.DeleteFiles,
+		snapshotFilesFn:           fs.SnapshotFiles,
+		sleepFn:                   time.Sleep,
+		identifierPool:            opts.IdentifierPool(),
+		contextPool:               opts.ContextPool(),
+		flushState:                newShardFlushState(),
+		tickWg:                    &sync.WaitGroup{},
+		coldWritesEnabled:         namespaceMetadata.Options().ColdWritesEnabled(),
+		emptySnapshotOnDiskByTime: make(map[xtime.UnixNano]bool),
```

Comment:
As discussed, let's move `emptySnapshotOnDiskByTime` to `shard.flushState` which already has a lock then we can use that lock to encapsulate the field to avoid the write lock in snapshot?